File: app/src/main/java/com/lasercyber/lws/ui/bean/http/CameraShowTimeRequest.java
package com.lasercyber.lws.ui.bean.http;

import com.google.gson.annotations.SerializedName;

import java.io.Serializable;
import java.util.Calendar;
import java.util.TimeZone;

import lombok.Data;
import lombok.experimental.Accessors;

/** JSON body for camera {@code PUT /System/showtime}. */
@Accessors(chain = true)
@Data
public class CameraShowTimeRequest implements Serializable {
    private Integer enable;
    private Integer positionx;
    private Integer positiony;
    private Integer year;

    @SerializedName("mon")
    private Integer month;

    private Integer day;
    private Integer hour;

    @SerializedName("min")
    private Integer minute;

    private Integer sec;

    public static CameraShowTimeRequest create(int enable, int positionX, int positionY, boolean useUtcNow) {
        CameraShowTimeRequest request = new CameraShowTimeRequest()
                .setEnable(enable)
                .setPositionx(positionX)
                .setPositiony(positionY);
        if (useUtcNow) {
            fillUtcNow(request);
        } else {
            fillZeros(request);
        }
        return request;
    }

    public static void fillUtcNow(CameraShowTimeRequest request) {
        Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        request.setYear(calendar.get(Calendar.YEAR));
        request.setMonth(calendar.get(Calendar.MONTH) + 1);
        request.setDay(calendar.get(Calendar.DAY_OF_MONTH));
        request.setHour(calendar.get(Calendar.HOUR_OF_DAY));
        request.setMinute(calendar.get(Calendar.MINUTE));
        request.setSec(calendar.get(Calendar.SECOND));
    }

    public static void fillZeros(CameraShowTimeRequest request) {
        request.setYear(0);
        request.setMonth(0);
        request.setDay(0);
        request.setHour(0);
        request.setMinute(0);
        request.setSec(0);
    }
}

File: app/lws_hmi/lib/features/ip_camera/application/camera_show_time_request.dart
/// JSON body for camera `PUT /System/showtime` (lws-ui `CameraShowTimeRequest`).
///
/// When [create] fills the clock (`fillNow: true`), fields use the device
/// **local** wall clock so OSD DateTimeOverlay matches HMI timezone (CST etc.).
/// lws-ui historically filled UTC here; that made the burned-in clock look wrong.
final class CameraShowTimeRequest {
  const CameraShowTimeRequest({
    required this.enable,
    required this.positionx,
    required this.positiony,
    required this.year,
    required this.month,
    required this.day,
    required this.hour,
    required this.minute,
    required this.sec,
  });

  final int enable;
  final int positionx;
  final int positiony;
  final int year;

  /// Camera API field `mon`.
  final int month;
  final int day;
  final int hour;

  /// Camera API field `min`.
  final int minute;
  final int sec;

  factory CameraShowTimeRequest.create({
    required int enable,
    required int positionX,
    required int positionY,
    required bool fillNow,
    DateTime? now,
  }) {
    final base = CameraShowTimeRequest(
      enable: enable,
      positionx: positionX,
      positiony: positionY,
      year: 0,
      month: 0,
      day: 0,
      hour: 0,
      minute: 0,
      sec: 0,
    );
    if (!fillNow) {
      return base;
    }
    final t = now ?? DateTime.now();
    return CameraShowTimeRequest(
      enable: enable,
      positionx: positionX,
      positiony: positionY,
      year: t.year,
      month: t.month,
      day: t.day,
      hour: t.hour,
      minute: t.minute,
      sec: t.second,
    );
  }

  Map<String, Object?> toJson() => <String, Object?>{
        'enable': enable,
        'positionx': positionx,
        'positiony': positiony,
        'year': year,
        'mon': month,
        'day': day,
        'hour': hour,
        'min': minute,
        'sec': sec,
      };
}
